forest/benchmarking/superoperator_tools.py

```diff
@@ -363,3 +367,105 @@ def computational2pauli_basis_matrix(dim) -> np.ndarray:
+
+
+#TODO Honest approximations for Channels that act on one or MORE qubits.
+
+# ==================================================================================================
+# Apply channel
+# ==================================================================================================
+def apply_kraus_ops_2_state(kraus_ops: Sequence[np.ndarray], state: np.ndarray) -> np.ndarray:
+    r"""
+    Apply a quantum channel, specified by Kraus operators, to state. The Kraus operators need not be
+    square.
+
+    :param kraus_ops: A list or tuple of N Kraus operators, each operator is M by dim ndarray
+    :param state_matrix: A dim by dim ndarray
+    :return: M by M ndarray
+    """
+    if isinstance(kraus_ops, np.ndarray):  # handle input of single kraus op
+        if len(kraus_ops[0].shape) < 2:  # first elem is not a matrix
+            kraus_ops = [kraus_ops]
+
+    dim,_ = state.shape
+    rows,cols = kraus_ops[0].shape
+
+    if rows==cols:
+        # square Kraus operators
+        if dim != rows:
+            raise ValueError("Dimensions of state and Kraus operator are incompatible")
+
+    if rows!=cols:
+        # Non-square Kraus operators
+        if dim != cols:
+            raise ValueError("Dimensions of state and Kraus operator are incompatible")
+
+    new_state = np.zeros((rows,rows))
+    for M in kraus_ops:
+        new_state += M @ state @ np.transpose(M.conj())
+
+    return new_state
+
+def apply_choi_matrix_2_state(choi: np.ndarray, state: np.ndarray) -> np.ndarray:
+    r"""
+    Apply a quantum channel, specified by a Choi matrix (using the column stacking convention),
+    to a state.
+
+    The Choi matrix is a dim**2 by dim**2 matrix and the state rho is a dim by dim matrix. The
+    output state is
+
+    rho_{out} = Tr_{A}[(rho^T \otimes Id) Choi_matrix ],
+
+    where Tr_{A} is the partial trace over hilbert space H_A with respect to the tensor product
+    H_A \otimes H_B, and T denotes transposition.
+
+
+    :param choi: a dim**2 by dim**2 matrix
+    :return: a dim by dim matrix.
+    """
+    dim = int(np.sqrt(np.asarray(choi).shape[0]))
+    dims = [dim, dim]
+    Id = np.identity(dim)
+    tot_matrix = np.kron(state.transpose(),Id) @ choi
+    return partial_trace(tot_matrix, [1], dims)
+
+def apply_lioville_matrix_2_state(superoperator: np.ndarray, state: np.ndarray) -> np.ndarray:
+    r"""
+    Apply a quantum channel, specified by a superoperator aka the Lioville or Pauli-Lioville
+    representation, to a state by matrix multiplication.
+
+    Note: you must ensure the state and superoperator are represented in the same basis.
+
+    :param choi: a dim**2 by dim**2 matrix.
+    :return: a dim by dim matrix.
+    """
+    dim = int(np.sqrt(np.asarray(superoperator).shape[0]))
+    rows, cols = state.shape
+
+    if dim**2 != rows:
+        raise ValueError("Dimensions of the vectorized state and superoperator are incompatible")
+    return superoperator@state
```

Comment:
Do we really need this one? The operation is trivial and numpy already provides dimension checking.

Additionally apply_choi_matrix_2_state takes in a density matrix not a vectorized state. I think this ambiguity is enough reason to not include this method--it would be easier to do the multiplication myself than to look up which argument it takes.